File: services/executor/internal/service/executor.go
package service

import (
	"context"
	"errors"
	"fmt"
	"os/exec"

	"github.com/myntdeveloper/executor/internal/models"
	"github.com/myntdeveloper/executor/internal/runner"
)

type ExecuteService struct {
}

func NewExecuteService() *ExecuteService {
	return &ExecuteService{}
}

func (s *ExecuteService) Execute(language, code, input string) (models.ExecuteResponse, error) {
	runner := runner.DockerRunner{}
	stdout, stderr, err := runner.Run(language, code, input)
	timeout := false
	if err != nil {
		var execErr *exec.Error
		if errors.As(err, &execErr) && execErr.Err == exec.ErrNotFound {
			return models.ExecuteResponse{}, fmt.Errorf("docker is not installed or not available in PATH %s", err)
		}
		if errors.Is(err, context.DeadlineExceeded) {
			timeout = true
		} else {
			if stderr == "" {
				stderr = err.Error()
			}
			return models.ExecuteResponse{
				Output:  stdout,
				Error:   stderr,
				Timeout: false,
			}, nil
		}
	}

	return models.ExecuteResponse{
		Output:  stdout,
		Error:   stderr,
		Timeout: timeout,
	}, nil
}

File: frontend/src/app/api/executor.ts
import { apiFetch } from "./client";

interface ExecuteResponse {
  output?: string;
  error?: string;
  stderr?: string;
  stdout?: string;
  timeout?: boolean;
}

export const execute = async (language: string, code: string, input: string) => {
  const res = await apiFetch("/executor/execute", {
    method: "POST",
    body: JSON.stringify({
      language,
      code,
      input,
    }),
  });

  if (!res.ok) {
    const errorText = await res.text();
    try {
      const errorBody = JSON.parse(errorText) as { error?: string };
      throw new Error(errorBody.error || "Execution failed");
    } catch {
      throw new Error(errorText || "Execution failed");
    }
  }

  return (await res.json()) as ExecuteResponse;
};
